packit/api.py

```diff
@@ -211,6 +213,122 @@ def update_dist_git(
+                )
+        return patch
+
+    def update_source_git(self, revision_range: str):
+        """Update a source-git repo from a dist-git repo
+
+        Synchronizes the spec file and commits in the given revision range.
+        The Version tag must not have changed and the sources and patches
+        in dist-git must not have been touched.
+
+        Args:
+            revision_range: Range of commits from dist-git to convert to
+                source-git. Specified in git-log-like format.
+
+        Raises:
+            PackitException: If the given update cannot be performed, i.e.
+                sources or patches were touched or Version changed.
+        """
+        if self.dg.specfile.get_version() != self.up.specfile.get_version():
+            raise PackitException(
+                "The version in the spec file of the provided dist-git is "
+                "different from the source-git one. Rebasing is currently "
+                "not supported."
+            )
+
+        dg_release = self.dg.specfile.get_release()
+        up_release = self.up.specfile.get_release()
+        if dg_release != up_release:
+            logger.info(
+                f"Release differs between dist-git and source-git ("
+                f"{dg_release} in dist-git and {up_release} in source-git). "
+                f"Versions are the same, continuing with the update."
+            )
+
+        # Do the checks beforehand but store commits and diffs to avoid recomputing.
+        # Getting patch of a git commit is costly as per GitPython docs.
+        commits: List[git.Commit] = []
+        diffs: List[List[git.Diff]] = []
+        patch_suffix = ".patch"
+        distro_path = Path(self.up.local_project.working_dir) / DISTRO_DIR
+        for commit in self.dg.local_project.get_commits(revision_range, reverse=True):
+            commits.append(commit)
+            diffs.append(self.dg.local_project.get_commit_diff(commit))
+            for diff in diffs[-1]:
+                if diff.a_path == "sources" or diff.b_path == "sources":
+                    raise PackitException(
+                        f"The sources file was modified in commit "
+                        f"{commit.hexsha} which is part of the provided range. "
+                        f"Such operation is not supported."
+                    )
+                a_path = diff.a_path or ""
+                b_path = diff.b_path or ""
+                # FIXME: this check is not great, but if we want to be more precise, we would
+                #   have to parse the spec in each checkout of dist-git
+                if a_path.endswith(patch_suffix) or b_path.endswith(patch_suffix):
+                    raise PackitException(
+                        f"A patch was modified in commit {commit.hexsha} "
+                        f"which is not supported by this command."
+                    )
+
+        for i, commit in enumerate(commits):
```

Comment:
Would be nice to have an info-log telling the number of commits that are going to be synced.